src/configuration/services/varnish.md

```diff
@@ -0,0 +1,114 @@
+
+If you have multiple applications fronted by the same Varnish instance then you will need to include logic to determine to which application a request is forwarded.  For example:
+
+```
+sub vcl_recv {
+    if (req.url ~ "^/foo/") {
+        set req.backend_hint = app.backend();
+    } else {
+        set req.backend_hint = app2.backend();
+    }
+}
+```
+
+This configuration will direct all requests to a URL beginning with a `/foo/` path to the application on the relationship `app`, and all other requests to the application on the relationship `app2`.
+
+Besides that, the VCL file, including the `vcl_recv()` function, can be arbitrarily complex to suit the needs of the project.  See the [Varnish documentation](https://varnish-cache.org/docs/index.html) for more details on the functionality offered by Varnish.
+
+> **note**
+> A misconfigured VCL file can result in incorrect, often mysterious and confusing behavior.  Platform.sh does not provide support for VCL configuration options beyond the basic connection logic documented here.
+
+### Route incoming requests to Varnish
+
+To enable Varnish now, edit the `.platform/routes.yaml` file to point to the Varnish service you just created.  You also need to disable the router cache as it is now entirely redundant with Varnish.
+
+For example:
+
+```yaml
+"https://{default}/":
+    type: upstream
+    upstream: "varnish:http"
+    cache: 
+        enabled: false
+```
+
+That will map all incoming requests to the Varnish service rather than the application.  Varnish will then, based on the VCL file, forward requests to the application as appropriate.
+
+## Circular relationships
+
+At this time Platform.sh does not support circular relationships between a service and an application.  That means you cannot add a relationship in your `.platform.app.yaml` that points to the Varnish service.  If you do so then one of the relationships will be skipped and the connection will not work.  This limitation may be lifted in the future.
+
+## Stats endpoint
+
+The Varnish service also offers an `http+stats` endpoint, which provides access to some Varnish analysis and debugging tools.  To access it, from a dedicated app container add the following to `.platform.app.yaml`:
+
+```yaml
+relationships:
+    varnishstats: `varnish:http+stats`
+```
+
+You can then access the `varnishstats` relationship over HTTP at the following paths to get diagnostic information:
+
+* `/`: returns the error if generating the VCL failed with an error
+* `/config`: returns the generated VCL
+* `/stats`: returns the output of `varnishstat`
+* `/logs`: returns a streaming response of `varnishlog`
+
+Note that because of the circular relationship issue noted above this cannot be done on the application that Varnish is forwarding to.  It will need to be run on a separate application container.  As a result it is most useful for debugging in development environments, not in production.
```

Comment:
I'd skip the last sentence.

Reply:
Also, this should be "this cannot be done on _an_ application that Varnish is forwarding to", no?